README.md
# Node.js HTTP Server Mock

Mockable http server for your functional tests.

`npm install mock-http-server`


### Full working example

```js
var ServerMock = require("mock-http-server");

describe('Test', function() {

    // Run an HTTP server on localhost:9000
    var server = new ServerMock({ host: "localhost", port: 9000 });

    beforeEach(function(done) {
        server.start(done);
    });

    afterEach(function(done) {
        server.stop(done);
    });

    it('should do something', function(done) {
        server.on({
            method: 'GET',
            path: '/resource',
            reply: {
                status:  200,
                headers: { "content-type": "application/json" },
                body:    JSON.stringify({ hello: "world" })
            }
        });

        // Now the server mock will handle a GET http://localhost:9000/resource
        // and will reply with 200 `{"hello": "world"}`
        done();
    });
});
```


### Methods

#### Constructor

`new ServerMock(httpConfig, httpsConfig)` instance a new mockable HTTP/HTTPS Server. If `httpConfig` is defined, creates an HTTP server, while if `httpsConfig` is defined, creates an HTTPS server. They can be both defined.

Example:
```js
var server = new ServerMock({
    host: "localhost",
    port: 80
}, {
    host: "localhost",
    port: 443,
    key: fs.readFileSync("private-key.pem"),
    cert: fs.readFileSync("certificate.pem")
});
```


#### `start(callback)`

Starts the server and invokes the callback once ready to accept connections.

Example:
```js
beforeEach(function(done) {
    server.start(done);
});
```


#### `stop(callback)`

Stops the server and invokes the callback all resources have been released.

Example:
```js
afterEach(function(done) {
    server.stop(done);
});
```

#### `on(options)`

Defines a request handler. Multiple calls to `on()` can be chained together.

| Option                   | Default                                  | Description |
| ------------------------ | ---------------------------------------- | ----------- |
| `method`                 | `GET`                                    | HTTP method to match. Can be `*` to match any method. |
| `path`                   |                                          | HTTP request path to match. Can be `*` to match any path (to be used in conjunction with filter to allow custom matching)|
| `filter`                 |                                          | The value is a filter function `fn(request)`: if it returns `true` the handler gets executed. |
| `reply.status`           | `200`                                    | HTTP response status code. Can be a `number` or a synchronous function `fn(request)` that returns the response status code. |
| `reply.headers`          | `{ "content-type": "application/json" }` | HTTP response headers. `content-length` is managed by the server implementation. |
| `reply.headersOverrides` | `{ "content-length": 1000 }`             | HTTP response headers to override to default headers (ie. `content-length`). If a value is set to `undefined`, the header gets removed from the response. |
| `reply.body`             | empty string                             | HTTP response body. Can be a `string`, a synchronous function `fn(request)` that returns the body, or an asynchronous function `fn(request, reply)` that send the response body invoking `reply(body)`. |
| `reply.end`              | `true`                                   | End the response once the body has been sent (default behaviour). If `false`, it will keep the response connection open indefinitely (useful to test special cases on the client side - ie. read timeout after partial body response sent). |
| `delay`                  | 0                                        | Delays the response by X milliseconds. |


Example:
```js
server.on({
    method: 'GET',
    path: '/resource',
    reply: {
        status:  200,
        headers: { "content-type": "application/json" },
        body:    JSON.stringify({ hello: "world" })
    }
});
```

or:
```js
server.on({
    method: '*',
    path: '/resource',
    reply: {
        status:  200,
        headers: { "content-type": "application/json" },
        body:    function(req) {
            return req.method === "GET" ? JSON.stringify({ action: "read" }) : JSON.stringify({ action: "edit" });
        }
    }
});
```

or:
```js
server.on({
    method: '*',
    path: '/resource',
    reply: {
        status:  200,
        headers: { "content-type": "application/json" },
        body:    function(req, reply) {
            setTimeout(function() {
                reply(req.method === "GET" ? JSON.stringify({ action: "read" }) : JSON.stringify({ action: "edit" }));
            }, 100);
        }
    }
});
```

or (JSON is parsed when the Content-Type is 'application/json'):
```js
server.on({
    method: 'POST',
    path: '/resources',
    filter: function (req) {
      return _.isEqual(req.body, {
        name: 'someName',
        someOtherValue: 1234
      })
    },
    reply: {
        status:  201,
        headers: { "content-type": "application/json" },
        body: {
          id: 987654321,
          name: 'someName',
          someOtherValue: 1234
        }
    }
});
```


#### `requests(filter)`

Returns an array containing all requests received. If `filter` is defined, it allows to filter requests by `method`, `path`, or both.

| Filter          | Description |
| --------------- | ----------- |
| `method`        | Filter requests by method. |
| `path`          | Filter requests by path. |

Example:
```js
// Returns all requests
server.requests();

// Returns all GET requests
server.requests({ method: "GET" });

// Returns all GET requests to /resource
server.requests({ method: "GET", path: "/resource" });
```


#### `connections()`

Returns an array containing all active connections.


#### `getHttpPort()`

Returns the port at which the HTTP server is listening to or `null` otherwise. This may be useful if you configure the HTTP server port to `0` which means the operating system will assign an arbitrary unused port.


#### `getHttpsPort()`

Returns the port at which the HTTPS server is listening to or `null` otherwise. This may be useful if you configure the HTTPS server port to `0` which means the operating system will assign an arbitrary unused port.

#### `reset()`

Clears request handlers and requests received by the HTTP server.

#### `resetHandlers()`

Clears all request handlers that were previously set using `on()` method.


#### `resetRequests()`

Clears all requests received by the HTTP server.



## Contribute

### How to publish a new version

**Release npm package**:

1. Update version in `package.json` and `package-lock.json`
2. Update `CHANGES.md`
3. Release new version on GitHub
4. Run `npm publish`



### License

MIT


CHANGES.md
# Changelog

#### 1.4.5 (2021-11-17)
 * Fixed `multipart/form-data` double body parse error.
#### 1.4.4 (2021-06-08)
 * Improve fallback to `text/plain` body parse in the request when format is not handled.
#### 1.4.3 (2021-05-28)
 * Fallback to `text/plain` body parse in the request when format is not handled.
#### 1.4.1 (2019-10-07)
 * Fixed `requests()` filtering: multiple conditions are now in **and** instead of **or** (PR [#21](https://github.com/spreaker/node-mock-http-server/pull/21) - thanks to [danroshko](https://github.com/danroshko))

#### 1.4.0 (2019-09-16)
 * Added support to parse request body with `Content-Type: text-plain` (PR [#20](https://github.com/spreaker/node-mock-http-server/pull/20) - thanks to [esanchezm](https://github.com/esanchezm))

#### 1.3.0 (2019-07-31)
 * Added `resetHandlers()` (PR [#19](https://github.com/spreaker/node-mock-http-server/pull/19) - thanks to [jyte](https://github.com/jyte))

#### 1.2.0 (2018-12-12)
 * Added `getHttpPort()` and `getHttpsPort()` (PR [#18](https://github.com/spreaker/node-mock-http-server/pull/18) - thanks to [Nautigsam](https://github.com/Nautigsam))

#### 1.1.0 (2018-11-14)
 * Added support to parse urlencoded request body (PR [#9](https://github.com/spreaker/node-mock-http-server/pull/9) - thanks to [zhaotian2470](https://github.com/zhaotian2470))
 * Removed obsolete code in the parsing of JSON-encoded request body (PR [#16](https://github.com/spreaker/node-mock-http-server/pull/16) - thanks to [pitpit](https://github.com/pitpit))

#### 1.0.0 (2018-04-26)
 * Added `reply.end` support
 * Added `connections()` support
 * Documented `reply.headersOverrides`
 * Remove headers from response, whose value is `undefined`
 * End all the connections on `server.stop()` before closing the HTTP server

#### 0.2.0 (2017-10-18)
 * Added wildcard support to `path` (thanks to [danedmunds](https://github.com/danedmunds))
 * Parse JSON in incoming request body when request content-type is "application/json" (thanks to [danedmunds](https://github.com/danedmunds))

#### 0.1.0 (2017-02-03)
 * Handlers are now appended to the begin of the list, when registering. This allows to override handlers in subsequent `on()` calls (see [Issue #2](https://github.com/spreaker/node-mock-http-server/issues/2)).

#### 0.0.4 (2016-02-04)
 * FIX: clear requests array on `server.stop()`


spec/helpers/client.js
const http = require("http");


/**
 * Sends an HTTP request to the server and returns a Promise which resolve
 * with the HTTP response containing the body too.
 *
 * @param  {String} host
 * @param  {String} port
 * @param  {String} method
 * @param  {String} path
 * @return {Promise}
 */
module.exports.request = (host, port, method, path, headers, body) => {
    return new Promise((resolve, reject) => {
        const req = http.request({ host: host, port: port, method: method, headers: headers, path: path }, (res) => {
            res.body = "";

            res.on("data", (chunk) => {
                res.body += chunk;
            });

            res.on("end", () => {
                resolve(res);
            });
        });

        req.on("error", (err) => {
            reject(err);
        });

        if (body) {
            req.write(body);
        }

        req.end();
    });
};


spec/http_server_spec.js
const ServerMock = require("../src/server.js");
const client     = require("./helpers/client.js");


describe("ServerMock", () => {
    let server;

    afterEach((done) => {
        if (!server) {
            return done();
        }

        server.stop(() => {
            server = null;
            done();
        });
    });


    describe("getHttpPort()", () => {
        it("should return the HTTP server port specified in the config at which the server is listening to", (done) => {
            server = new ServerMock({ host: "localhost", port: 21256 });
            expect(server.getHttpPort()).toBeNull();

            server.start(() => {
                expect(server.getHttpPort()).toBe(21256);
                done();
            });
        });

        it("should return the HTTP server port autoassigned by the operating system if set to zero in the config", (done) => {
            server = new ServerMock({ host: "localhost", port: 0 });
            expect(server.getHttpPort()).toBeNull();

            server.start(() => {
                expect(server.getHttpPort()).not.toBeNull();
                expect(server.getHttpPort()).not.toBe(0);
                done();
            });
        });
    });


    describe("on()", () => {
        beforeEach((done) => {
            server = new ServerMock({ host: "localhost", port: 0 });
            server.start(done);
        });

        it("should mock the server behaviour for the configure method and path", async () => {
            server.on({
                method: "GET",
                path:   "/resource",
                reply:  {
                    status:  200,
                    headers: { "content-type": "application/json" },
                    body:    JSON.stringify({ hello: "world" })
                }
            });

            const res = await client.request("localhost", server.getHttpPort(), "GET", "/resource");

            expect(res.statusCode).toBe(200);
            expect(res.headers["content-type"]).toBe("application/json");
            expect(res.body).toBe('{"hello":"world"}');
        });

        it("should return 404 is the requested method and path has not been mocked", async () => {
            const res = await client.request("localhost", server.getHttpPort(), "GET", "/resource");

            expect(res.statusCode).toBe(404);
            expect(res.headers["content-type"]).toBe("plain/text");
            expect(res.body).toBe("Not Found");
        });

        it("should get a text plain body when content-type is text/plain", async () => {
            server.on({
                method: "POST",
                path: "/resource",
                reply: {
                    status: function(req) {
                        if (req.body !== "Hello world\nThis is a text") {
                            return 403;
                        }
                        return 200;
                    }
                }
            });

            const res = await client.request("localhost", server.getHttpPort(), "POST", "/resource", { "Content-Type": "text/plain" }, "Hello world\nThis is a text");

            expect(res.statusCode).toBe(200);
        });

        it("should get a text plain body as a fallback", async () => {
            server.on({
                method: "POST",
                path: "/resource",
                reply: {
                    status: function(req) {
                        if (req.body !== "{}\n{}\n") {
                            return 403;
                        }
                        return 200;
                    }
                }
            });

            const res = await client.request("localhost", server.getHttpPort(), "POST", "/resource",  { "Content-Type": "application/x-ndjson" }, "{}\n{}\n");

            expect(res.statusCode).toBe(200);
        });

        it("should get a text plain body as a fallback when no content type is passed", async () => {
            server.on({
                method: "POST",
                path: "/resource",
                reply: {
                    status: function(req) {
                        if (req.body !== "{}\n{}\n") {
                            return 403;
                        }
                        return 200;
                    }
                }
            });

            const res = await client.request("localhost", server.getHttpPort(), "POST", "/resource",  {}, "{}\n{}\n");

            expect(res.statusCode).toBe(200);
        });

        it("should successfully parse a multipart form", async () => {
            server.on({
                method: "POST",
                path: "/resource",
                reply: {
                    status: 204,
                    headers: {},
                    body: null,
                }
            });

            const res = await client.request("localhost", server.getHttpPort(), "POST", "/resource", {
                'content-type': 'multipart/form-data; boundary=--------MYBOUNDARY',
            }, [
                "--------MYBOUNDARY",
                "Content-Disposition: form-data; name=\"hello\"",
                "",
                "world!",
            ].join("\n"));

            expect(res.statusCode).toBe(204);
        });
    });

    
    describe("requests()", () => {
        beforeEach(done => {
            server = new ServerMock({ host: "localhost", port: 0 });
            server.start(async () => {
                server.on({
                    method: "PUT",
                    path:   "/dog",
                    reply:  {
                        status:  200,
                        headers: { "content-type": "application/json" },
                        body:    JSON.stringify({ breed: "German Shepherd" })
                    }
                });
                server.on({
                    method: "POST",
                    path:   "/cat",
                    reply:  {
                        status:  200,
                        headers: { "content-type": "application/json" },
                        body:    JSON.stringify({ breed: "Maine coon" })
                    }
                });
                server.on({
                    method: "DELETE",
                    path:   "/bird",
                    reply:  {
                        status:  204,
                        headers: { "content-type": "application/json" }
                    }
                });

                await client.request("localhost", server.getHttpPort(), "PUT", "/dog", { "Content-Type": "application/json" }, JSON.stringify({ breed: "German Shepherd" })).catch(error => console.error(error));
                await client.request("localhost", server.getHttpPort(), "POST", "/cat", { "Content-Type": "application/json" }, JSON.stringify({ breed: "Maine coon" })).catch(error => console.error(error));
                await client.request("localhost", server.getHttpPort(), "DELETE", "/bird").catch(error => console.error(error));

                done();
            });
        });

        it("should return all requests when given no arguments", async() => {
            const requests = server.requests();

            expect(requests).not.toBeNull();
            expect(requests.length).toBe(3);
        });
        
        it("should return the request emitted on dog API when path equals 'dog'", () => {
            const requests = server.requests({ path: '/dog' });

            expect(requests).not.toBeNull();
            expect(requests.length).toBe(1);
            expect(requests[0].pathname).toBe('/dog');
            expect(requests[0].method).toBe('PUT');
        });
                
        it("should return the request emitted on cat API when method equals 'POST'", () => {
            const requests = server.requests({ method: 'POST' });

            expect(requests).not.toBeNull();
            expect(requests.length).toBe(1);
            expect(requests[0].pathname).toBe('/cat');
            expect(requests[0].method).toBe('POST');
        });
                        
        it("should return the request emitted on bird API when method equals 'DELETE' and path equals 'bird'", () => {
            const requests = server.requests({ path: '/bird', method: 'DELETE' });

            expect(requests).not.toBeNull();
            expect(requests.length).toBe(1);
            expect(requests[0].pathname).toBe('/bird');
            expect(requests[0].method).toBe('DELETE');
        });
    });

    describe("reset()", () => {
        beforeEach(done => {
            server = new ServerMock({ host: "localhost", port: 0 });
            server.start(async () => {
                server.on({
                    method: "GET",
                    path:   "/resource",
                    reply:  {
                        status:  200,
                        headers: { "content-type": "application/json" },
                        body:    JSON.stringify({ hello: "world" })
                    }
                });

                await client.request("localhost", server.getHttpPort(), "GET", "/resource");

                done();
            });
        });

        it("should return only Not Found error after being invoked", async () => {
            const res1 = await client.request("localhost", server.getHttpPort(), "GET", "/resource");
            expect(res1.statusCode).toBe(200);

            server.reset();

            // As there isn't any handler anymore, returns a Not Found error.
            const res2 = await client.request("localhost", server.getHttpPort(), "GET", "/resource");
            expect(res2.statusCode).toBe(404);
        });

        it("should leave an empty requests array", () => {
            server.reset();

            const requests = server.requests();
            expect(requests).not.toBeNull();

            expect(requests.length).toBe(0, "Requests weren't cleaned");
        });
    });

    describe("resetRequests()", () => {
        beforeEach(done => {
            server = new ServerMock({ host: "localhost", port: 0 });
            server.start(async () => {
                server.on({
                    method: "GET",
                    path:   "/resource",
                    reply:  {
                        status:  200,
                        headers: { "content-type": "application/json" },
                        body:    JSON.stringify({ hello: "world" })
                    }
                });

                await client.request("localhost", server.getHttpPort(), "GET", "/resource");

                done();
            });
        });

        it("should leave an empty requests array", () => {
            server.resetRequests();

            const requests = server.requests();
            expect(requests).not.toBeNull();

            expect(requests.length).toBe(0, "Requests weren't cleaned");
        });
    });
    
    describe("resetHandlers()", () => {
        beforeEach(done => {
            server = new ServerMock({ host: "localhost", port: 0 });
            server.start(() => {
                server.on({
                    method: "GET",
                    path:   "/resource",
                    reply:  {
                        status:  200,
                        headers: { "content-type": "application/json" },
                        body:    JSON.stringify({ hello: "world" })
                    }
                });

                done();
            });
        });

        it("should return only Not Found error after being invoked", async () => {
            const res1 = await client.request("localhost", server.getHttpPort(), "GET", "/resource");
            expect(res1.statusCode).toBe(200);

            server.resetHandlers();

            // As there isn't any handler anymore, returns a Not Found error.
            const res2 = await client.request("localhost", server.getHttpPort(), "GET", "/resource");
            expect(res2.statusCode).toBe(404);
        });
    });
});


src/server.js
var connect     = require('connect'),
    url         = require('url'),
    fs          = require('fs'),
    http        = require('http'),
    https       = require('https'),
    util        = require('util'),
    _           = require('underscore'),
    multiparty  = require('multiparty'),
    bodyParser  = require('body-parser');

/**
 * @param {String} host     Server host
 * @param {String} port     Server port
 * @param {String} key      HTTPS key (if missing the server is HTTP)
 * @param {String} cert     HTTPS certificate (if missing the server is HTTP)
 */
function Server(host, port, key, cert)
{
    var server      = null,
        address     = null,
        handlers    = [],
        requests    = [],
        connections = [],
        self = this;

    function _saveRequest(req, res, next) {
        requests.push(req);

        next();
    }

    function _multipart(req, res, next) {

        if (req.method !== 'POST' && req.method !== 'PUT') {
            return next();
        }

        if ('multipart/form-data' !== (req.headers['content-type'] || '').split(';')[0]) {
            return next();
        }

        var form = new multiparty.Form();

        form.parse(req, function(err, fields, files) {

            // mark the body as parsed so other parsers don't attempt to parse the request
            req._body = true;

            req.body  = {};
            req.files = {};

            if (err) { return next(); }

            _.each(fields, function(value, name){
                if (Array.isArray(value) && value.length === 1) {
                    req.body[name] = value[0];
                } else {
                    req.body[name] = value;
                }
            });

            _.each(files, function(value, name){
                if (Array.isArray(value) && value.length === 1) {
                    req.files[name] = value[0];
                } else {
                    req.files[name] = value;
                }
            });

            next();
        });
    }

    /**
     * Supports:
     * - { reply: { body: "data" }}
     * - { reply: { body: function(req) { return "data"; }}}
     * - { reply: { body: function(req, send) { setTimeout(function() { send("data"); }, 1000); }}}
     */
    function _getResponseBody(handler, req, callback) {
        // String
        if (!_(handler.reply.body).isFunction()) {
            return callback(handler.reply.body || "");
        }

        // Synch function
        if (handler.reply.body.length <= 1) {
            return callback(handler.reply.body(req) || "");
        }

        // Asynch function
        handler.reply.body(req, callback);
    }

    /**
     * Supports:
     * - { reply: { status: 200 }}
     * - { reply: { status: function(req) { return 200; }}}
     */
    function _getResponseStatus(handler, req, callback) {
        // Number
        if (!_(handler.reply.status).isFunction()) {
            return callback(handler.reply.status || 0);
        }

        // Synch function
        return callback(handler.reply.status(req) || 0);
    }

    function _handleMockedRequest(req, res, next) {

        var handled = false;

        _(handlers).each(function(handler) {

            // Parse request URL
            var reqParts = url.parse(req.url, true);
            req.pathname = reqParts.pathname;
            req.query    = reqParts.query;

            // Check if we can handle the request
            if (handled ||
              (handler.method != "*" && req.method != handler.method.toUpperCase()) ||
              (handler.path != "*" && reqParts.pathname != handler.path) ||
              (handler.filter && handler.filter(req) !== true)) {
                return;
            }

            // Flag as handled
            handled = true;

            // Get response status and body
            _getResponseBody(handler, req, function(content) {
                _getResponseStatus(handler, req, function(status) {

                    // Prepare response data
                    var encoding = Buffer.isBuffer(content) ? undefined : "utf8";
                    var length   = Buffer.isBuffer(content) ? content.length : Buffer.byteLength(content, "utf8");

                    // Prepare headers
                    var headersToSend = {};
                    var headers = _(handler.reply.headers)
                        .extend({ "content-length": length }, handler.reply.headersOverrides || {});

                    // Remove undefined values from headers (useful to remove content-length from the response if
                    // overridden with undefined)
                    _.each(headers, function(headerValue, headerName) {
                        if (headerValue === undefined) {
                            delete headers[headerName];
                        }
                    });

                    // Remove "null" values from headers
                    _(headers).each(function(value, name) {
                        if (value !== null) {
                            headersToSend[name] = value;
                        }
                    });

                    // Send response (supports delay)
                    setTimeout(function() {

                        // Send headers
                        res.writeHead(status, headersToSend);

                        // Send content
                        if (req.method != "HEAD") {
                            res.write(content, encoding);
                        }

                        // End response (unless explicitly stated to keep it open)
                        if (handler.reply.end !== false) {
                            res.end();
                        }

                    }, handler.delay || 0);

                });
            });
        });

        if (!handled) {
            next();
        }
    }

    function _handleDefaultRequest(req, res, next)
    {
        res.writeHead(404, {
            "content-type":   "plain/text",
            "content-length": 9
        });

        res.end("Not Found");
    }

    this.start = function(callback)
    {
        // Create app stack
        var connectApp = connect()
            .use(_saveRequest)
            .use(_multipart)
            .use(bodyParser.json())
            .use(bodyParser.text())
            .use(bodyParser.urlencoded({extended: true}))
            .use(bodyParser.text({type: function() { return true; }}))
            .use(_handleMockedRequest)
            .use(_handleDefaultRequest);

        // Create server
        if (key && cert) {
            server = https.createServer({key: key, cert: cert}, connectApp);
        } else {
            server = http.createServer(connectApp);
        }

        server.on("connection", function (connection) {
            connection.on("close", function () {
                connections = _(connections).without(connection);
            });

            connections.push(connection);
        });

        server.on("listening", function () {
            address = server.address();
            callback();
        });

        server.listen(port, host);
    };


    this.stop = function (callback) {
        if (!server) {
            return callback();
        }

        // Close connections
        _(connections).forEach(function (connection) {
            connection.end();
        });

        server.on("close", function() {
            server   = null;
            address  = null;
            handlers = [];

            // Wait until all connections are closed
            if (connections.length === 0) {
                requests = [];

                callback();
            } else {
                _(connections).forEach(function (connection) {
                    connection.on("close", function() {
                        connections = _(connections).without(connection);
                        if (connections.length === 0) {
                            requests = [];

                            callback();
                        }
                    });
                });
            }
        });

        server.close();
    };

    this.on = function(handler) {
        // Add default reply
        handler.reply         = _({}).extend({ "status": 200, "body": "" }, handler.reply);
        handler.reply.headers = _({}).extend({ "content-type": "application/json" }, handler.reply.headers);

        // Add default method
        handler = _({}).extend({ "method": "GET" }, handler);

        handlers.unshift(handler);
        return this;
    };

    /**
     * Clears request handlers and requests received by the HTTP server.
     */
    this.reset = function() {
        self.resetHandlers();
        self.resetRequests();
    }

    /**
     * Clears all request handlers that were previously set using `on()` method.
     */
    this.resetHandlers = function() {
        handlers = [];
    };

    /**
     * Clears all requests received by the HTTP server.
     */
    this.resetRequests = function() {
        requests = [];
    }

    /**
     * Returns an array containing all requests received. If `filter` is defined,
     * filters the requests by:
     * - method
     * - path
     *
     * @param  {Object} filter
     * @return {Array}
     */
    this.requests = function(filter) {
        return _(requests).filter(function(req) {
            var reqParts = url.parse(req.url, true);
            const methodMatch = !filter || !filter.method || filter.method === req.method;
            const pathMatch = !filter || !filter.path || filter.path === reqParts.pathname;

            return methodMatch && pathMatch;
        });
    };

    /**
     * Returns an array containing all active connections.
     *
     * @return {Array}
     */
    this.connections = function() {
        return connections;
    };

    /**
     * Returns the port number if set or null otherwise
     *
     * @return {Number|null}
     */
    this.getPort = function() {
        return address ? address.port : null;
    }
}

function ServerVoid() {

    this.on            = function() {};
    this.start         = function(callback) { callback(); };
    this.stop          = function(callback) { callback(); };
    this.requests      = function() { return []; };
    this.connections   = function() { return []; };
    this.getPort       = function() { return null; };
    this.reset         = function() {};
    this.resetHandlers = function() {};
    this.resetRequests = function() {};
}

/**
 * @param {Object} httpConfig
 * @param {Object} httpsConfig
 */
function ServerMock(httpConfig, httpsConfig)
{
    var httpServerMock  = httpConfig ?  new Server(httpConfig.host, httpConfig.port) : new ServerVoid();
    var httpsServerMock = httpsConfig ? new Server(httpsConfig.host, httpsConfig.port, httpsConfig.key, httpsConfig.cert) : new ServerVoid();

    this.start = function(callback) {
        httpServerMock.start(function() {
            httpsServerMock.start(function() {
                callback();
            });
        });
    };

    this.stop = function(callback) {
        httpServerMock.stop(function() {
            httpsServerMock.stop(callback);
        });
    };

    this.on = function(handler) {
        httpServerMock.on(handler);
        httpsServerMock.on(handler);

        return this;
    };

    this.requests = function(filter) {
        return httpServerMock.requests(filter).concat(httpsServerMock.requests(filter));
    }

    this.connections = function() {
        return httpServerMock.connections().concat(httpsServerMock.connections());
    }

    this.getHttpPort = function() {
        return httpServerMock.getPort();
    }

    this.getHttpsPort = function() {
        return httpsServerMock.getPort();
    }

    this.reset = function() {
        httpServerMock.reset();
        httpsServerMock.reset();
    }

    this.resetHandlers = function() {
        httpServerMock.resetHandlers();
        httpsServerMock.resetHandlers();
    }

    this.resetRequests = function() {
        httpServerMock.resetRequests();
        httpsServerMock.resetRequests();
    }
}


module.exports = ServerMock;
